packages/office-ui-fabric-react/src/utilities/selection/Selection.ts

```diff
@@ -135,37 +135,43 @@ export class Selection implements ISelection {
       }
     }
 
-    if (shouldClear) {
+    if (shouldClear || items.length === 0) {
       this.setAllSelected(false);
     }
 
     // Check the exemption list for discrepencies.
     const newExemptedIndicies: { [key: string]: boolean } = {};
+    let newExemptedCount = 0;
 
     for (const indexProperty in this._exemptedIndices) {
-      if (this._exemptedIndices.hasOwnProperty(indexProperty)) {
+      if (Object.prototype.hasOwnProperty.call(this._exemptedIndices, indexProperty)) {
         const index = Number(indexProperty);
         const item = this._items[index];
         const exemptKey = item ? this.getKey(item, Number(index)) : undefined;
         const newIndex = exemptKey ? newKeyToIndexMap[exemptKey] : index;
 
         if (newIndex === undefined) {
-          // We don't know the index of the item any more so it's either moved or removed.
-          // In this case we reset the entire selection.
-          this.setAllSelected(false);
-          break;
+          // The item has likely been replaced or removed.
+          hasSelectionChanged = true;
```

Comment:
If we couldn't find the index, why no longer reset? Why remove comments?

Reply:
Resetting was inconsistent with the 'select all' case. What if some items from what you have selected have been removed? Why were we just clearing the selection as a result?

Reply:
It's been a while, but I believe the edge case is this:

Selection can't differentiate between a deletion vs a "cache hasn't downloaded that item" scenario.

In a random access model or even a paging model, you download a page of items, and then download the next page and find out something has changed, what do you do?

If you have a "change" api which gives you the changes since the last fetch, you can patch page 1. Or, you could repair page 1 by re-downloading all pages up to the current known truth. 

In a random access model, you might clear page 1 because page 2 is the only truth you know. In this case, any items from page 1 are in a "not downloaded yet" state, and selection can't differentiate between that and removed state. So the best effort safest behavior is to reset the selection.

Maybe this is too aggressive, or maybe the scenario is no longer valid. If you select all in a folder, where items are being moved/deleted asynchronously, and you get an push notification of changes which remove the item, selection resets would be annoying. But also, if we don't reset, you'd instead want your selection count to decrease, so maybe the ideal default behavior is to simply "unselect" missing items. Or have an opt in mode to do that, because unselecting in the case of "item is just not materialized in the dataset because of a cache reset" would be terrible for the user. 

E.g. select an item, scroll to page 2, reset and remove first item from cache, select an item from page 2, hit delete. Only 1 item gets deleted.